File: 3Dify/3Dify/Parallax/Encoders/ParallaxOcclusion/ParallaxOcclusionPassEncoder.m

#import "ParallaxOcclusionPassEncoder.h"
#import <simd/simd.h>

typedef struct {
    simd_float1 offsetX;
    simd_float1 offsetY;
    simd_float1 focalPoint;
} ParallaxOcclusionUniforms;

@interface ParallaxOcclusionPassEncoder ()
@property (nonatomic, strong) id<MTLDevice> device;
@property (nonatomic, strong) id<MTLRenderPipelineState> pipelineState;
@property (nonatomic, strong) id<MTLBuffer> parallaxOcclusionUniformsBuffer;
@property (nonatomic) simd_float1 offsetX;
@property (nonatomic) simd_float1 offsetY;
@property (nonatomic) simd_float1 focalPoint;
@end

@implementation ParallaxOcclusionPassEncoder

-(instancetype)initWithDevice:(id<MTLDevice>)device
{
    self = [super init];
    if (self) {
        self.device = device;
        self.pipelineState = [self parallaxOcclusionPipelineStateOnDevice:device];
        self.parallaxOcclusionUniformsBuffer = [self makeParallaxOcclusionUniformsBuffer];
        self.offsetX = 0.0f;
        self.offsetY = 0.0f;
        self.focalPoint = 0.0f;
    }
    return self;
}

-(id<MTLBuffer>)makeParallaxOcclusionUniformsBuffer
{
    id<MTLBuffer> buffer = [self.device newBufferWithLength:sizeof(ParallaxOcclusionUniforms)
                                                    options:MTLResourceOptionCPUCacheModeDefault];
    buffer.label = @"Parallax Occlusion Uniforms";
    return buffer;
}

-(id<MTLRenderPipelineState>)parallaxOcclusionPipelineStateOnDevice:(id<MTLDevice>)device
{
    id<MTLLibrary> library = [device newDefaultLibrary];
    MTLRenderPipelineDescriptor *descriptor = [MTLRenderPipelineDescriptor new];
    descriptor.label = @"Parallax Occlusion Pipeline State";
    descriptor.vertexFunction = [library newFunctionWithName:@"project_texture"];
    descriptor.fragmentFunction = [library newFunctionWithName:@"parallax_occlusion"];
    descriptor.colorAttachments[0].pixelFormat = MTLPixelFormatRGBA16Float;
    descriptor.colorAttachments[1].pixelFormat = MTLPixelFormatRGBA16Float;
    descriptor.depthAttachmentPixelFormat = MTLPixelFormatInvalid;
    NSError *error = nil;
    id<MTLRenderPipelineState> pipelineState = [device newRenderPipelineStateWithDescriptor:descriptor error:&error];
    if (!pipelineState) {
        NSLog(@"Error occurred when creating render pipeline state: %@", error);
    }
    return pipelineState;
}

-(void)  encodeIn:(id<MTLCommandBuffer>)commandBuffer
inputColorTexture:(id<MTLTexture>)inputColorTexture
inputDepthTexture:(id<MTLTexture>)inputDepthTexture
outputColorTexture:(id<MTLTexture>)outputColorTexture
outputDepthTexture:(id<MTLTexture>)outputDepthTexture
     drawableSize:(CGSize)drawableSize
       clearColor:(MTLClearColor)clearColor
{
    MTLRenderPassDescriptor* descriptor = [self outputToColorTextureDescriptorOfSize:drawableSize
                                  clearColor:clearColor
                                  outputColorTexture:outputColorTexture
                                  outputDepthTexture:outputDepthTexture];
    [self updateParallaxOcclusionUniforms];
    id<MTLRenderCommandEncoder> encoder = [commandBuffer renderCommandEncoderWithDescriptor:descriptor];
    [encoder setLabel:@"Parallax Occlusion"];
    [encoder setRenderPipelineState:self.pipelineState];
    [encoder setFragmentTexture:inputColorTexture atIndex:0];
    [encoder setFragmentTexture:inputDepthTexture atIndex:1];
    [encoder setFragmentBuffer:self.parallaxOcclusionUniformsBuffer offset:0 atIndex:0];
    [encoder drawPrimitives:MTLPrimitiveTypeTriangleStrip vertexStart:0 vertexCount:4];
    [encoder endEncoding];
}

-(void)updateOffsetX:(simd_float1)offsetX
          andOffsetY:(simd_float1)offsetY
{
    self.offsetX = offsetX;
    self.offsetY = offsetY;
}

-(void)updateFocalPoint:(simd_float1)focalPoint
{
    self.focalPoint = focalPoint;
}

-(void)updateParallaxOcclusionUniforms
{
    ParallaxOcclusionUniforms uniforms = (ParallaxOcclusionUniforms) {
        .offsetX = self.offsetX,
        .offsetY = self.offsetY,
        .focalPoint = self.focalPoint
    };
    memcpy(self.parallaxOcclusionUniformsBuffer.contents, &uniforms, sizeof(ParallaxOcclusionUniforms));
}

-(MTLRenderPassDescriptor *)outputToColorTextureDescriptorOfSize:(CGSize)size
                                                      clearColor:(MTLClearColor)clearColor
                                              outputColorTexture:(id<MTLTexture>)outputColorTexture
                                              outputDepthTexture:(id<MTLTexture>)outputDepthTexture
{
    MTLRenderPassDescriptor *descriptor = [MTLRenderPassDescriptor renderPassDescriptor];
    descriptor.colorAttachments[0].texture = outputColorTexture;
    descriptor.colorAttachments[0].clearColor = clearColor;
    descriptor.colorAttachments[0].loadAction = MTLLoadActionLoad;
    descriptor.colorAttachments[0].storeAction = MTLStoreActionStore;
    descriptor.colorAttachments[1].texture = outputDepthTexture;
    descriptor.colorAttachments[1].clearColor = clearColor;
    descriptor.colorAttachments[1].loadAction = MTLLoadActionLoad;
    descriptor.colorAttachments[1].storeAction = MTLStoreActionStore;
    descriptor.renderTargetWidth = size.width;
    descriptor.renderTargetHeight = size.height;
    return descriptor;
}

@end

File: ThreeDify/ThreeDify/Parallax/Encoders/ParallaxOcclusionPassEncoder.swift
import Foundation
import MetalKit
import simd

struct ParallaxOcclusionUniforms {
    var offsetX: simd_float1
    var offsetY: simd_float1
    var focalPoint: simd_float1
    
    static var initial: Self {
        .init(offsetX: 0, offsetY: 0, focalPoint: 0)
    }
}

struct ParallaxOcclusionPassEncoder {
    private let device: MTLDevice
    private let pipelineState: MTLRenderPipelineState
    private let uniformsBuffer: MTLBuffer
    
    var uniforms: ParallaxOcclusionUniforms
    
    enum InitializationError: Error {
        case libraryInitFailed
        case pipelineInitFailed
        case bufferInitFailed
    }
    
    enum RenderingError: Error {
        case makeDescriptorFailed
    }
    
    init(device: MTLDevice, uniforms: ParallaxOcclusionUniforms = .initial) throws {
        self.device = device
        
        guard
            let library = device.makeDefaultLibrary()
        else { throw InitializationError.libraryInitFailed }
        
        let descriptor = MTLRenderPipelineDescriptor()
        descriptor.label = "Parallax Occlusion Pipeline State"
        descriptor.vertexFunction = library.makeFunction(name: "project_texture")
        descriptor.fragmentFunction = library.makeFunction(name: "parallax_occlusion")
        descriptor.colorAttachments[0].pixelFormat = .rgba16Float
        descriptor.colorAttachments[1].pixelFormat = .rgba16Float
        descriptor.depthAttachmentPixelFormat = .invalid
        
        guard
            let pipelineState = try? device.makeRenderPipelineState(
                descriptor: descriptor
            )
        else { throw InitializationError.pipelineInitFailed }
        self.pipelineState = pipelineState
        guard
            let buffer = device.makeBuffer(
                length: MemoryLayout.size(ofValue: uniforms),
                options: .cpuCacheModeWriteCombined
            )
        else { throw InitializationError.bufferInitFailed }
        buffer.label = "Parallax Occlusion Uniforms Buffer"
        self.uniformsBuffer = buffer
        
        self.uniforms = uniforms
    }
    
    mutating func encode(
        inCommandBuffer commandBuffer: MTLCommandBuffer,
        inputColorTexture: MTLTexture,
        inputDepthTexture: MTLTexture,
        outputColorTexture: MTLTexture,
        outputDepthTexture: MTLTexture,
        drawableSize: CGSize,
        clearColor: MTLClearColor
    ) throws {
        let descriptor = MTLRenderPassDescriptor()
        descriptor.colorAttachments[0].texture = outputColorTexture
        descriptor.colorAttachments[0].clearColor = clearColor
        descriptor.colorAttachments[0].loadAction = .load
        descriptor.colorAttachments[0].storeAction = .store
        descriptor.colorAttachments[1].texture = outputDepthTexture
        descriptor.colorAttachments[1].clearColor = clearColor
        descriptor.colorAttachments[1].loadAction = .load
        descriptor.colorAttachments[1].storeAction = .store
        descriptor.renderTargetWidth = Int(drawableSize.width)
        descriptor.renderTargetHeight = Int(drawableSize.height)
        memcpy(
            uniformsBuffer.contents(),
            &uniforms,
            MemoryLayout.size(ofValue: uniforms)
        )
        guard
            let encoder = commandBuffer.makeRenderCommandEncoder(
                descriptor: descriptor
            )
        else { throw RenderingError.makeDescriptorFailed }
        encoder.label = "Parallax Occlusion"
        encoder.setRenderPipelineState(pipelineState)
        encoder.setFragmentTexture(inputColorTexture, index: 0)
        encoder.setFragmentTexture(inputDepthTexture, index: 1)
        encoder.setFragmentBuffer(uniformsBuffer, offset: 0, index: 0)
        encoder.drawPrimitives(type: .triangleStrip, vertexStart: 0, vertexCount: 4)
        encoder.endEncoding()
    }
}
